### PR TITLE
Reduce amount of code in headers

### LINKED ISSUE #57: Too much code in headers
Can we move all function definitions into cpp files so that it is easier to understand the roles of the various classes?

### PR DESCRIPTION
**Pull request by LB--**
Fixes #57, like #59 was originally intended to do.

There wasn't really that much that could be done anyway, but I did what I could. I also fixed the inconsistent include guard naming, which I think was my fault to begin with. The compile speed seems slightly faster on my machine, mostly because I moved some includes from headers to source files to reduce the amount of redundant work the compiler has to do. Also, I found some places where we could use `<iosfwd>` instead of `<istream>` or `<iostream>`, which should help too.

With `JsonReader` I switched to the PImpl idiom, but I was able to get away with just forward declaring the `_json_value` struct and using pointers/references to it in the header instead of full-fledged smart pointers and private impl structs.

### CODE REVIEW

File: src/piece/Pawn.hpp [hunk 1]
```diff
@@ -8,7 +8,8 @@ namespace chesspp
 {
     namespace piece
     {
-        class Pawn : public virtual Piece
+        class Pawn
```
Comment (ResidentBiscuit) [inline src/piece/Pawn.hpp]: What was your rationale for this change? I don't think we ever inherit from more than one class, so the line shouldn't be getting long and this goes against every code base I've seen. Not that I'm totally against it, I just don't think it makes anything cleaner or messier.

File: src/piece/Piece.hpp [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+#ifndef chesspp_piece_Piece_HeaderPlusPlus
+#define chesspp_piece_Piece_HeaderPlusPlus
 
 #include "config/BoardConfig.hpp"
 
 #include <memory>
 #include <set>
 #include <typeinfo>
-#include <iostream>
+#include <iosfwd>
```
Comment (ResidentBiscuit) [inline src/piece/Piece.hpp]: I never knew this header existed, that's pretty cool. +1

File: src/piece/Queen.hpp [hunk 3]
```diff
@@ -1,5 +1,5 @@
-#ifndef ChessPlusPlus_Piece_QueenChessPiece_HeaderPlusPlus
-#define ChessPlusPlus_Piece_QueenChessPiece_HeaderPlusPlus
+#ifndef chesspp_piece_Queen_HeaderPlusPlus
+#define chesspp_piece_Queen_HeaderPlusPlus
```
Comment (ResidentBiscuit) [inline src/piece/Queen.hpp]: I see you reworked the include guards a little bit. Definitely a lot cleaner and less annoying to write out. Make sure this gets documented in the style guide, if it's not already.

File: src/res/SfmlFileResource.hpp [hunk 4]
```diff
@@ -1,5 +1,5 @@
-#ifndef ChessPlusPlus_Res_SfmlFileResourceClass_HeaderPlusPlus
-#define ChessPlusPlus_Res_SfmlFileResourceClass_HeaderPlusPlus
+#ifndef chesspp_res_SfmlFileResource_HeaderPlusPlus
+#define chesspp_res_SfmlFileResource_HeaderPlusPlus
 
 #include "ResourceManager.hpp"
 //#include "SFML.hpp"
```
Comment (ResidentBiscuit) [inline src/res/SfmlFileResource.hpp]: I know it's part of this PR, but we should definitely remove commented out code (includes)

File: src/Debug.cpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                            return ch;
+                    }
+                    return traits_type::eof();
+                }
+                int sync() override
+                {
+                    return timestamp_and_flush() ? 0 : -1;
+                }
+
+                //returns std::string containing current system time.
```
Comment (ResidentBiscuit) [inline src/Debug.cpp]: Also not part of this PR, but all functions should be documented in Doxygen format.

File: src/piece/Pawn.hpp (hunk 1 above)
Comment (LB--) [inline src/piece/Pawn.hpp]: It's in the styleguide that it should be that way.

File: src/piece/Queen.hpp (hunk 3 above)
Comment (LB--) [inline src/piece/Queen.hpp]: Will do.

File: src/res/SfmlFileResource.hpp (hunk 4 above)
Comment (LB--) [inline src/res/SfmlFileResource.hpp]: It's kind of a technicality - the template only works with SFML classes but it doesn't need the include until it is instantiated. What do you suggest?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -34,6 +34,7 @@ Style Guide & Coding Conventions
 - Try to avoid excess or repetitive "access:" specifiers.
 - Avoid strange spaces before open parens and open braces and in other places unless you are lining up code with other code for readability and emphasis of relation.
 - There is no line character limit, but that is because you shouldn't have exceedingly long lines (unless they just have exceedingly long identifiers). Instead, avoid deeply nested compound expressions - they reduce readability and hinder understanding of what code is for and/or doing.
+- Include guards should mirror the primary namespace + class of the file, e.g. `chesspp::piece::Piece` maps to `chesspp_piece_Piece_HeaderPlusPlus`
```
Comment (LB--) [inline CONTRIBUTING.md]: @ResidentBiscuit is this fine?

Comment (ResidentBiscuit) [inline CONTRIBUTING.md]: Looks good to me. Just good to have a line in there to keep people consistent.

File: src/piece/Pawn.hpp (hunk 1 above)
Comment (ResidentBiscuit) [inline src/piece/Pawn.hpp]: Hmm closest thing I saw was for initializer lists.

Comment (LB--) [inline src/piece/Pawn.hpp]: It's supposed to apply to both, do you think the wording should be modified?

Comment (ResidentBiscuit) [inline src/piece/Pawn.hpp]: Yeah I say modify it or add a new entry for inheritance. It wasn't obvious to me that it was meant for both situations, which means it'll catch someone else also.

Comment (ResidentBiscuit) [inline src/piece/Pawn.hpp]: +1

**Comment by ResidentBiscuit** [comment]
I think everything's been addressed. Squash and merge when you're ready.
